source/common/router/rds_impl.cc

```diff
@@ -210,6 +221,34 @@ void RdsRouteConfigProviderImpl::onConfigUpdate() {
     prev_config->config_ = new_config;
     return previous;
   });
+
+  const auto aliases = config_update_info_->aliasesInLastVhdsUpdate();
+  // Regular (non-VHDS) RDS updates don't populate aliases fields in resources.
+  if (aliases.empty()) {
+    return;
+  }
+
+  // Notifies connections that RouteConfiguration update has been propagated.
+  // Callbacks processing is performed in FIFO order. The callback is skipped if alias(es) used in
+  // the VHDS update request do not match the aliases in the update response
+  config_update_callbacks_->runOnAllThreads(
+      [aliases](ThreadLocal::ThreadLocalObjectSharedPtr previous)
+          -> ThreadLocal::ThreadLocalObjectSharedPtr {
+        auto callbacks = std::dynamic_pointer_cast<ThreadLocalCallbacks>(previous)->callbacks_;
+        std::vector<std::string> aliases_not_in_update;
+        for (auto it = callbacks.begin(); it != callbacks.end();) {
+          std::set_difference(it->aliases_.begin(), it->aliases_.end(), aliases.begin(),
+                              aliases.end(), std::back_inserter(aliases_not_in_update));
+          if (aliases_not_in_update.empty()) {
+            it->cb_();
```

Comment:
I don't know why some of my previous comments disappeared.
but I think there is a life-cycle problem here: the cb_ owner (i.e., the activeStreams) may be dead when the on-demand xDS  response comes back, have a weak_ptr of the activeStream (or some similar mechanism) to detect if the activeStream is still alive is desired.
A RDS subscription could live longer than a listener/HCM with @lambdai 's recent update to RDS.

Reply:
Please see the latest changes, I think I have sorted out lifecycle issues.